modules/core/src/main/scala/org/tessellation/infrastructure/dag/DAGService.scala

```diff
@@ -15,23 +19,44 @@ object DAGService {
   def make[F[_]: Async](globalSnapshotStorage: GlobalSnapshotStorage[F]): DAGService[F] =
     new DAGService[F] {
 
-      def getBalance(address: Address): F[Balance] =
-        globalSnapshotStorage.head.map {
-          case Some(snapshot) => snapshot.value.info.balances.get(address).getOrElse(Balance.empty)
-          case None           => Balance.empty
+      def getBalance(address: Address): F[Option[(Balance, SnapshotOrdinal)]] =
+        globalSnapshotStorage.head.flatMap {
+          case Some(snapshot) => getBalance(snapshot.value.ordinal, address)
+          case _              => none.pure[F]
         }
 
-      def getTotalSupply: F[BigInt] = {
+      def getBalance(ordinal: SnapshotOrdinal, address: Address): F[Option[(Balance, SnapshotOrdinal)]] =
+        globalSnapshotStorage.get(ordinal).map {
+          case Some(snapshot) =>
+            val balance = snapshot.value.info.balances.get(address).getOrElse(Balance.empty)
+            val ordinal = snapshot.value.ordinal
+
+            (balance, ordinal).some
+          case _ => none
```

Comment:
`Option` makes the fallback to `None` implicitly by design so if you need to do it manually then you probably shouldn't pattern match there. It is like `if (none) then none`.

I would consider using `map` then but it looks ugly with `maybeSnapshot`:

```scala
globalSnapshotStorage.get(ordinal).map { maybeSnapshot =>
  snapshot.map(...)
}
```

or would consider temporal monad transformer:

```scala
OptionT(globalSnapshotStorage.get(ordinal))
  .map { snapshot =>
     val balance = ...
     val ordinal = ...
     (balance, ordinal)
  }.value
```